### PR TITLE
Refactor register search and support all search API filter operations

### LINKED ISSUE #2209: Refactor register search to use SDK Search API to perform database query
**Describe the feature request.**
Refactor register search to use SDK search DSL APIs instead of applying filter on the paginated items via code (very slow with a lot of data). 

**Additional context**
The current search implementation works by loading all the register data via the paging3 library, then applying the[ paging3 filter data transform function](https://developer.android.com/topic/libraries/architecture/paging/v3-transform#filter-data) to filter the data. 
_How the current filter works:_
1. A list of rule names is provided via the [RegisterConfiguration.searchBar.computedRules](https://github.com/opensrp/fhircore/blob/414d97d20310da9f8ad14ef0b00cba5975f2dcd3/android/engine/src/main/java/org/smartregister/fhircore/engine/configuration/register/RegisterConfiguration.kt#L34) 
2. For each of the item above extract the value from the computed values map of every register list item and compare it against the search term
3. Return a filtered list of ResourceData matching the condition above
4. Cache all the retrieved resource data for future use when performing search (subsequent search will be faster at the expense of more memory usage via the cache). Cache is cleared when the user exists the register.

PS: When searching register data, the result is not paginated but uses an infinite scroll which is not consistent with the figma design that uses pagination.

**Acceptance criteria**
Register search should return paginated search response when a user searches the register data. The search data queries should also be configurable.

**Area path**
A list of ordered steps in the app on usage of the feature to support anyone testing it e.g. Code reviewer, QA e.g.
1. Load household register for instance (or any register) 
2. Enter name of household to search in the search input box

**Implementation plan (For Engineers)**
- Provide search data queries via RegisterConfiguration
- On search term change, query the database using the above query configs
- Refactor `loadRegisterData` to query the database using the configured queries and finally paginate and display the result

Query? Do we want to support code level filters for search parameters that are not supported via the Search API? is it possible to make non-searchable params searchable via the SDK? e.g. if `Group.name` is not searchable can the `name` property be indexed to allow search?

### PR DESCRIPTION
**Pull request by ellykits**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2209  or Closes #2209 

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt [hunk 1]
```diff
@@ -181,5 +181,6 @@ fun String.castToType(type: DataType): Type? {
     DataType.URI -> UriType(this)
     DataType.CODING -> this.tryDecodeJson<Coding>()
     DataType.QUANTITY -> this.tryDecodeJson<Quantity>()
+    else -> null /*TODO cast the (several) remaining Enumeration.DataTypes*/
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:184]: is null OK or should we throw here?

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/SearchExtension.kt [hunk 2]
```diff
@@ -47,81 +48,76 @@ fun Search.filterByResourceTypeId(
...
 
 fun Search.filterBy(dataQuery: DataQuery) {
-  when (dataQuery.filterType) {
-    Enumerations.SearchParamType.TOKEN -> filterToken(dataQuery)
-    Enumerations.SearchParamType.STRING -> filterString(dataQuery)
-    Enumerations.SearchParamType.DATE -> filterDate(dataQuery)
-    else ->
-      throw UnsupportedOperationException("Can not apply ${dataQuery.filterType} as search filter")
+  dataQuery.filterCriteria.forEach { filterCriterion: FilterCriterionConfig ->
+    when (filterCriterion) {
```
Comment (pld) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/SearchExtension.kt:52]: do the conditions cover the possible types of `filterCriterion`? if not, is it better to ignore the filter (as these changes do) or throw?

**Review by pld — COMMENTED** [review]
looks great!

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/SearchExtension.kt (hunk 2 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/SearchExtension.kt:52]: The conditions cover all the possible types.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt (hunk 1 above)
Comment (ellykits) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/QuestionnaireExtension.kt:184]: We'd update the implementation here (maybe during CHT) to handle rest of the relevant DataTypes but for now returning `null` is okay.

**Review by pld — DISMISSED** [review]

**Review by pld — APPROVED** [review]
